File: packages/fintech/src/sbe.js
// sbe.js — schema-driven Simple Binary Encoding flyweight decoder.
//
// Schema descriptors are JSON (docs/adapters/MANAGED-SEAMS-V1.md §3). The
// golden fixture schema (sbe-schema.json) declares littleEndian: true; the
// decoder honors the declaration for every field read.
//
// Framing: records arrive as [u32 LE length][SBE message]. The decoder reads
// the 8-byte message header (blockLength, templateId, schemaId, version),
// resolves the template, decodes declared fixed fields, and SKIPS anything
// beyond blockLength (SBE extension fields) by the transport length —
// forward compatibility without schema churn.
//
// Zero-allocation contract: one decoder instance, one preallocated output
// vector; the caller's stable visitor callback receives (templateId, vector,
// fieldCount) with the SAME vector reused across records. No per-record
// objects.

export class SbeDecoder {
  constructor(schema) {
    this.schema = schema;
    this.littleEndian = schema.littleEndian !== false;
    this.blockLength = 0;
    this.templateId = 0;
    this.schemaId = 0;
    this.version = 0;
    this.truncated = 0;
    this.skipped = 0; // unknown templates
    this.bytesProcessed = 0;
    this._vec = new Float64Array(32); // reused output vector
  }

  #u16(dv, off) { return this.littleEndian ? dv.getUint16(off, true) : dv.getUint16(off, false); }
  #u32(dv, off) { return this.littleEndian ? dv.getUint32(off, true) : dv.getUint32(off, false); }
  #u64(dv, off) {
    if (this.littleEndian) {
      const lo = dv.getUint32(off, true);
      const hi = dv.getUint32(off + 4, true);
      return hi * 4294967296 + lo; // exact < 2^53
    }
    const hi = dv.getUint32(off, false);
    const lo = dv.getUint32(off + 4, false);
    return hi * 4294967296 + lo;
  }
  #i64(dv, off) {
    const raw = this.#u64(dv, off);
    // two's complement reinterpretation without BigInt
    return raw >= 9223372036854775808 ? raw - 18446744073709551616 : raw;
  }

  // Decode one record at `off` (after the transport length prefix).
  // Writes into the shared vector; returns the visitor arg triple via the
  // callback. Returns bytes consumed, or 0 on unknown template / short read.
  #decodeRecord(dv, off, visit) {
    if (off + 8 > dv.byteLength) { this.truncated++; return 0; }
    this.blockLength = this.#u16(dv, off);
    this.templateId = this.#u16(dv, off + 2);
    this.schemaId = this.#u16(dv, off + 4);
    this.version = this.#u16(dv, off + 6);
    const msg = this.schema.messages[this.templateId];
    if (!msg) { this.skipped++; return 0; }
    const fields = msg.fields;
    const vec = this._vec;
    const n = Math.min(fields.length, vec.length);
    for (let i = 0; i < n; i++) {
      const f = fields[i];
      const at = off + 8 + f.offset;
      switch (f.type) {
        case 'u8': vec[i] = dv.getUint8(at); break;
        case 'i8': vec[i] = dv.getInt8(at); break;
        case 'u16': vec[i] = this.#u16(dv, at); break;
        case 'i16': vec[i] = this.littleEndian ? dv.getInt16(at, true) : dv.getInt16(at, false); break;
        case 'u32': vec[i] = this.#u32(dv, at); break;
        case 'i32': vec[i] = this.littleEndian ? dv.getInt32(at, true) : dv.getInt32(at, false); break;
        case 'u64': case 'i64': vec[i] = f.type === 'u64' ? this.#u64(dv, at) : this.#i64(dv, at); break;
        case 'f32': vec[i] = this.littleEndian ? dv.getFloat32(at, true) : dv.getFloat32(at, false); break;
        case 'f64': vec[i] = this.littleEndian ? dv.getFloat64(at, true) : dv.getFloat64(at, false); break;
        default: vec[i] = 0;
      }
    }
    visit(this.templateId, vec, n);
    return 8 + this.blockLength;
  }

  // Walk the whole stream; visit(templateId, vector, fieldCount) per record.
  // Returns the number of records visited. `buffer` may be an ArrayBuffer,
  // a DataView, or a typed-array/Buffer view (byte offsets view-relative).
  process(buffer, byteLength, visit) {
    let dv;
    if (buffer instanceof DataView) {
      dv = buffer;
    } else if (buffer instanceof ArrayBuffer) {
      dv = new DataView(buffer, 0, byteLength);
    } else {
      dv = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    }
    let off = 0;
    let visited = 0;
    while (off + 4 <= byteLength) {
      const len = this.#u32(dv, off);
      if (off + 4 + len > byteLength) { this.truncated++; break; }
      const consumed = this.#decodeRecord(dv, off + 4, visit);
      if (consumed > 0) {
        visited++;
        this.bytesProcessed += 4 + len;
      }
      off += 4 + len;
    }
    return visited;
  }
}

File: python/weft_fintech/src/weft_fintech/sbe.py
# sbe.py — schema-driven SBE flyweight decoder (managed Python side).
#
# Mirrors packages/fintech/src/sbe.js: [u32 LE length][SBE message] framing,
# 8-byte message header, blockLength-gated fixed fields, extension bytes
# skipped by transport length, unknown templates counted and skipped.
# Field values land in a preallocated shared list (reused across records).

import struct

_U32 = struct.Struct('<I')
_U16 = struct.Struct('<H')

_READERS = {}


def _reader(type_name, little):
    e = '<' if little else '>'
    fn = {
        'u8': lambda mv, off: mv[off],
        'i8': lambda mv, off: struct.unpack_from(e + 'b', mv, off)[0],
        'u16': lambda mv, off: struct.unpack_from(e + 'H', mv, off)[0],
        'i16': lambda mv, off: struct.unpack_from(e + 'h', mv, off)[0],
        'u32': lambda mv, off: struct.unpack_from(e + 'I', mv, off)[0],
        'i32': lambda mv, off: struct.unpack_from(e + 'i', mv, off)[0],
        'u64': lambda mv, off: struct.unpack_from(e + 'Q', mv, off)[0],
        'i64': lambda mv, off: struct.unpack_from(e + 'q', mv, off)[0],
        'f32': lambda mv, off: struct.unpack_from(e + 'f', mv, off)[0],
        'f64': lambda mv, off: struct.unpack_from(e + 'd', mv, off)[0],
    }[type_name]
    return fn


class SbeDecoder:
    __slots__ = ('schema', 'little', 'block_length', 'template_id',
                 'schema_id', 'version', 'truncated', 'skipped',
                 'bytes_processed', '_vec', '_templates')

    def __init__(self, schema):
        self.schema = schema
        self.little = schema.get('littleEndian', True)
        self.block_length = 0
        self.template_id = 0
        self.schema_id = 0
        self.version = 0
        self.truncated = 0
        self.skipped = 0
        self.bytes_processed = 0
        self._vec = [0] * 32
        self._templates = {}
        for tid, msg in schema.get('messages', {}).items():
            fields = msg['fields']
            self._templates[int(tid)] = (
                msg['blockLength'],
                [(f['offset'], _reader(f['type'], self.little)) for f in fields],
            )

    def process(self, buffer, byte_length, visit):
        mv = memoryview(buffer)
        off = 0
        visited = 0
        vec = self._vec
        while off + 4 <= byte_length:
            (length,) = _U32.unpack_from(mv, off)
            if off + 4 + length > byte_length:
                self.truncated += 1
                break
            rec = off + 4
            if rec + 8 <= len(mv):
                self.block_length = _U16.unpack_from(mv, rec)[0]
                self.template_id = _U16.unpack_from(mv, rec + 2)[0]
                self.schema_id = _U16.unpack_from(mv, rec + 4)[0]
                self.version = _U16.unpack_from(mv, rec + 6)[0]
                tpl = self._templates.get(self.template_id)
                if tpl is None:
                    self.skipped += 1
                else:
                    block_length, fields = tpl
                    n = min(len(fields), len(vec))
                    for i in range(n):
                        foff, fn = fields[i]
                        vec[i] = fn(mv, rec + 8 + foff)
                    visit(self.template_id, vec, n)
                    visited += 1
                    self.bytes_processed += 4 + length
            off += 4 + length
        return visited
